external/fv3fit/fv3fit/_shared/taper_function.py

```diff
@@ -0,0 +1,106 @@
+import abc
+import numpy as np
+import xarray as xr
+
+from fv3fit._shared.novelty_detector import NoveltyDetector
```

Comment:
Is this a circular import? It seems like `NoveltyDetector` is the parent object here...so the `.load` methods below probably shouldn't need access to it. It seems like they only use it to get default values. Maybe define the defaults in this module instead?

Reply:
I don't think this is a circular import, since the parent object is `OutOfSampleModel` (which also contains a `NoveltyDetector` object). If possible, I think I'd prefer to keep the default logic in the individual implementations of `NoveltyDetector`, since the default cutoffs are computed in very different ways. (e.g. minmax is always zero, OCSVM is a negative number determined by the maximum score of all training samples, and the ensemble approach (if implemented) will be a positive variance, likely also the maximum observed scores.). Since it isn't a circular import, is it fine if I have something like this, which populates a config file based on the default values and passes it to the other taper function? 

```
def get_taper_function(name: str = _MASK_NAME, config: dict = {}) -> Callable[[xr.DataArray], xr.DataArray]:
    if name == _MASK_NAME:
        return lambda x: taper_mask(x, **config)
    elif name == _RAMP_NAME:
        return lambda x: taper_ramp(x, **config)
    elif name == _DECAY_NAME:
        return lambda x: taper_decay(x, **config)
    else:
        raise ValueError("Incorect")


def get_taper_function_from_novelty_detector(novelty_detector: NoveltyDetector, name: str = _MASK_NAME) -> Callable[[xr.DataArray], xr.DataArray]:
    cutoff = novelty_detector._get_default_cutoff()
    config = {
        "cutoff": cutoff,
        "ramp_min": cutoff,
        "threshold": cutoff
    }
    if cutoff > 0:
        config["ramp_max"] = 2 * cutoff
    elif cutoff < 0:
        config["ramp_max"] = cutoff / 2
    else:
        config["ramp_max"] = 1
    return get_taper_function(name, config)
```

Reply:
Actually, nevermind, it might be easiest to do this in `OutOfSampleModel.load`. I'll mock something up in a commit in a bit.

Reply:
>  I think I'd prefer to keep the default logic in the individual implementations of NoveltyDetector, since the default cutoffs are computed in very different ways. (e.g. minmax is always zero, OCSVM is a negative number determined by the maximum score of all training samples, and the ensemble approach (if implemented) will be a positive variance, likely also the maximum observed scores.)

Interesting. It sounds like `NoveltyDetector` basically owns the concept of `TaperFunction` and could potentially hide this concept from other objects... in particular the `OutOfSampleModel`. Currently  `NoveltyDetector` and `OutOfSampleModel` and `TaperFunction` all know about tapering...which seems like a poor separation of concerns.

There are a couple of options to reduce this coupling:
1. Remove the concept of `OutOfSampleModel` and simply call the appropriate tapering logic in your `NoveltyDetector` implementations. 
2. Make `NoveltyDetector.predict_novelties` return a score  `alpha` from 0 to 1 that reflects the amount of tapering you want to apply. Then `OutOfSampleModel.predict` can use the score like this:
```
    def predict(self, X: xr.Dataset) -> xr.Dataset:
        base_predict = self.base_model.predict(X)
        # modify .precit_novelties to return a single dataarray of scores
        # score = 1 is OOS, score=0 is not.
        score = self.novelty_detector.predict_novelties(X, self.cutoff)
        masked_predict = base_predict.copy()
        for output_variable in self.base_model.output_variables:
            masked_predict[output_variable] =  (1-score) * base_predict[output_variable]
        return xr.merge([masked_predict, novelties])
```

Reply:
The second option is most direct generalization of what the code on master currently does IMO.

Reply:
I [talked](https://github.com/ai2cm/fv3net/pull/1945#discussion_r925891537) with @AnnaKwa and @oliverwm1 about using this kind of approach (and that's actually what was originally implemented in the first commit of the PR), but we were primarily thinking that tapering is something specific to the `OutOfSampleModel` rather than something intrinsic to a `NoveltyDetector`, since novelty detection alone isn't directly linked to the online predictions unless we specify that by creating an `OutOfSampleModel`. I'm ok with either way

Reply:
So one potential option would be to shift the score to make the default cutoff for each be zero, but I'm slightly hesitant to do that rn since each score is meaningful in the context of the novelty detector that produces it; even if the score is always negative for OCSVM and always non-negative for min-max, both are meaningful quantities that I would want to look at directly, given the context of the model. 

Also, there are already a bunch of saved runs with recorded novelty_scores as they're currently defined. To avoid inconsistency, I'd rather add a new field like `centered_score` than replace the old one.

Reply:
> since each score is meaningful in the context of the novelty detector that produces it

This makes sense. However, at the same time, we should hide this kind of detail from other components whenever possible to avoid coupling. I like the idea of adding a centered_score output. 

Perhaps you can modify `NoveltyDetector.predict_novelties` to return a tuple of (centered_score, diagnostics). centered_score will be a dataarray which means you can remove the reference to `SCORE_OUTPUT_VAR` (one less point of coupling). `diagnostics` is datasets/dict of dataarrays that can simply be merged into the output without inspecting its contents at all.

Reply:
> NoveltyDetector basically owns the concept of TaperFunction

I was kind of confused by this statement as I think the most recent iteration of the PR moved the tapering out of the NoveltyDetector. Although it is true that there's still some coupling between the tapering and novelty detectors via each detector's get_default_cutoff method. The suggestion to return the tuple of centered_score and diagnostics sounds like a good way to avoid that.

Reply:
This all sounds good! I'll add the centered score and have a new commit shortly.